### PR TITLE
Resolve issue CRPYT404 (Team 1.5)

### PR DESCRIPTION
**Pull request by AghilesAzzoug**
Resolve of issue [CRPYT404](https://github.com/green-code-initiative/ecoCode/issues/127).

# Summary
The use of generator comprehensions instead of list comprehensions in for-loops declaration can save RAM usage. It has multiple benefits like reducing CO2 emissions as well as releasing memory constraints on the hardware.

# Why it works?
Python generators resemble lazy lists from other programming languages: when iterated over, they compute their values on the fly. They lack some list behaviors (indexing, len method, ...) but are memory-efficient, as they do not store each of their values in memory, unlike lists. Thus, when declared in a for-loop declaration, list comprehensions can be safely replaced with generator comprehensions. 
For more details on list comprehensions vs generator comprehensions, see [Python documentation](https://docs.python.org/3/howto/functional.html#generator-expressions-and-list-comprehensions).

Local experiments* on list comprehensions vs generator comprehensions gives the following results on:

**1. Memory usage:**

memory_usage

**2. Co2 Emissions**

Using [CodeCarbon](https://codecarbon.io/) we get the following results:

co2_emission

For both metrics, the bigger the list, the greater is the gain is. 

*Specs for experimentations:

- 12th Gen Intel Core I7-12700H
- 16 Gb RAM (4800 Mhz)
- Windows 11 OS version 22H2
- Python 3.9.15
- memory_profiler==0.61.0 (for RAM experiments)
- codecarbon==2.1.4 (for CO2 emissions)

# On SonarQube 
The render of an issue:
image

The render of the issue explanation:
image

### CODE REVIEW

**Comment by dedece35** [comment]
Hi @AghilesAzzoug ,

please, could you :
- resolve conflicts
- take into account review notes
- add some tests in php-test-project

... in ordre to merge this PR.

File: python-plugin/src/main/java/fr/greencodeinitiative/python/checks/AvoidListComprehensionInIterations.java [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+                case "filter":
+                case "enumerate":
+                    Objects.requireNonNull(callExpression.argumentList()).
+                            arguments().forEach(e -> visitFunctionArguments(context, e.children().get(0)));
+            }
+        }
+    }
+
+    private void visitFunctionArguments(SubscriptionContext context, Tree argument) {
+        if (argument.is(LIST_COMPREHENSION)) {
```
Comment (dedece35) [inline python-plugin/src/main/java/fr/greencodeinitiative/python/checks/AvoidListComprehensionInIterations.java:51]: this is duplicated code with a part of previous function.
please make factorisation of this code.

File: python-plugin/src/main/java/fr/greencodeinitiative/python/checks/AvoidListComprehensionInIterations.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+            context.addIssue(argument.firstToken(), DESCRIPTION);
+
+        } else if (argument.is(CALL_EXPR)) {
+            CallExpression callExpression = (CallExpression) argument;
+            switch (callExpression.callee().firstToken().value()) {
+                case "zip":
```
Comment (dedece35) [inline python-plugin/src/main/java/fr/greencodeinitiative/python/checks/AvoidListComprehensionInIterations.java:57]: are you sure that this list is exhaustive ?

**Review by dedece35 — CHANGES_REQUESTED** [review]

File: python-plugin/src/main/java/fr/greencodeinitiative/python/checks/AvoidListComprehensionInIterations.java (hunk 1 above)
Comment (AghilesAzzoug) [reply python-plugin/src/main/java/fr/greencodeinitiative/python/checks/AvoidListComprehensionInIterations.java]: Done.

File: python-plugin/src/main/java/fr/greencodeinitiative/python/checks/AvoidListComprehensionInIterations.java (hunk 2 above)
Comment (AghilesAzzoug) [reply python-plugin/src/main/java/fr/greencodeinitiative/python/checks/AvoidListComprehensionInIterations.java]: We only focused on list comprehension in _iterations_ for this rule. We added support for native python functions (zip, filter, and enumerate) that are often used in iterations.

**Comment by AghilesAzzoug** [comment]
Hello @dedece35,

Thanks for reviewing this PR.

- Your review notes were taken into account.
- For the python-test-project, there's already a pending PR with test code for this issue (see [#1](https://github.com/green-code-initiative/ecoCode-python-test-project/pull/1)).
- CHANGELOG.md and RULES.md were updated.
- I changed some parts of the code to adhere to SonarQube conventions.

File: python-plugin/src/main/java/fr/greencodeinitiative/python/checks/AvoidListComprehensionInIterations.java [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+        tags = {"bug", "eco-design", "ecocode"})
+public class AvoidListComprehensionInIterations extends PythonSubscriptionCheck {
+
+    public static final String RULE_KEY = "EC404";
+    public static final String DESCRIPTION = "Use generator comprehension instead of list comprehension in for loop declaration";
+
+    @Override
+    public void initialize(Context context) {
+        context.registerSyntaxNodeConsumer(FOR_STMT, this::visitIteration);
+
```
Comment (dedece35) [inline python-plugin/src/main/java/fr/greencodeinitiative/python/checks/AvoidListComprehensionInIterations.java:35]: please remove this useless line :p

File: python-plugin/src/main/java/fr/greencodeinitiative/python/checks/AvoidListComprehensionInIterations.java [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+                break;
+        }
+    }
+
+    private void visitFunctionArgument(SubscriptionContext context, Tree argument) {
+        if (argument.is(REGULAR_ARGUMENT)) {
+            Expression expression = ((RegularArgument)argument).expression();
+            if (expression.is(LIST_COMPREHENSION)) {
+                context.addIssue(expression.firstToken(), DESCRIPTION);
+
```
Comment (dedece35) [inline python-plugin/src/main/java/fr/greencodeinitiative/python/checks/AvoidListComprehensionInIterations.java:70]: useless line

File: python-plugin/src/main/java/fr/greencodeinitiative/python/checks/AvoidListComprehensionInIterations.java [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+        context.registerSyntaxNodeConsumer(FOR_STMT, this::visitIteration);
+
+    }
+
+    private void visitIteration(SubscriptionContext context) {
+        ForStatement forStatement = (ForStatement) context.syntaxNode();
+
+        Expression forTestExpression = forStatement.testExpressions().get(0);
+        if (forTestExpression.is(LIST_COMPREHENSION)) {
+
```
Comment (dedece35) [inline python-plugin/src/main/java/fr/greencodeinitiative/python/checks/AvoidListComprehensionInIterations.java:43]: useless line

File: python-plugin/src/main/java/fr/greencodeinitiative/python/checks/AvoidListComprehensionInIterations.java [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+            context.addIssue(forTestExpression.firstToken(), DESCRIPTION);
+
```
Comment (dedece35) [inline python-plugin/src/main/java/fr/greencodeinitiative/python/checks/AvoidListComprehensionInIterations.java:45]: useless line

File: python-plugin/src/main/java/fr/greencodeinitiative/python/checks/AvoidListComprehensionInIterations.java [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+                context.addIssue(expression.firstToken(), DESCRIPTION);
+
+            } else if (expression.is(CALL_EXPR)) {
+                CallExpression callExpression = (CallExpression) expression;
+                visitCallExpression(context, callExpression);
+            }
+        }
+    }
+
+    private static String getFunctionNameFromCallExpression(CallExpression callExpression) {
```
Comment (dedece35) [inline python-plugin/src/main/java/fr/greencodeinitiative/python/checks/AvoidListComprehensionInIterations.java:78]: why static method ? regarding the content, why not but this method is only called from one non static method

**Review by dedece35 — CHANGES_REQUESTED** [review]

File: python-plugin/src/main/java/fr/greencodeinitiative/python/checks/AvoidListComprehensionInIterations.java (hunk 7 above)
Comment (AghilesAzzoug) [reply python-plugin/src/main/java/fr/greencodeinitiative/python/checks/AvoidListComprehensionInIterations.java]: Copy paste mistake, my bad :)

File: python-plugin/src/test/resources/checks/AvoidListComprehensionInIterations.py [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+
+def non_compliant_example_enumerate():
+    for idx, var in enumerate([var2 for var2 in range(1000)]): # Noncompliant {{Use generator comprehension instead of list comprehension in for loop declaration}}
+        print(var)
+
+def non_compliant_example_zip():
+    for var, var_ in zip([var2 for var2 in range(1000)], [var2 for var2 in range(1000)]): # Noncompliant {{Use generator comprehension instead of list comprehension in for loop declaration}} {{Use generator comprehension instead of list comprehension in for loop declaration}}
+        print(var)
+
+def non_compliant_example_zip():
```
Comment (dedece35) [inline python-plugin/src/test/resources/checks/AvoidListComprehensionInIterations.py:13]: there is two same methods named `non_compliant_example_zip`. why ?

**Review by dedece35 — CHANGES_REQUESTED** [review]

File: python-plugin/src/main/java/fr/greencodeinitiative/python/checks/AvoidListComprehensionInIterations.java [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+import static org.sonar.plugins.python.api.tree.Tree.Kind.FOR_STMT;
+import static org.sonar.plugins.python.api.tree.Tree.Kind.LIST_COMPREHENSION;
+import static org.sonar.plugins.python.api.tree.Tree.Kind.REGULAR_ARGUMENT;
+
+@Rule(
+        key = AvoidListComprehensionInIterations.RULE_KEY,
+        name = AvoidListComprehensionInIterations.DESCRIPTION,
+        description = AvoidListComprehensionInIterations.DESCRIPTION,
+        priority = Priority.MINOR,
+        tags = {"bug", "eco-design", "ecocode"})
```
[the remark this answers is not in the document]
Comment (AghilesAzzoug) [reply python-plugin/src/main/java/fr/greencodeinitiative/python/checks/AvoidListComprehensionInIterations.java:26]: True, that's definitely not a bug. But I saw this tag in all other python rules (which are not bugs also).

Should I remove the "bug" tag? Add an "optimization" one?

File: python-plugin/src/test/resources/checks/AvoidListComprehensionInIterations.py (hunk 8 above)
Comment (AghilesAzzoug) [reply python-plugin/src/test/resources/checks/AvoidListComprehensionInIterations.py]: Duplicated function removed. Same for `ecoCode-python-test-project`.

File: python-plugin/src/main/java/fr/greencodeinitiative/python/checks/AvoidListComprehensionInIterations.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (dedece35) [reply python-plugin/src/main/java/fr/greencodeinitiative/python/checks/AvoidListComprehensionInIterations.java:26]: for me, I should use SonarQube native existing tag `performance`
I agree with you : we have `bug` in all our rules ... but it's historic and not the good practice for me ... we have to refactor this `bug` tag.
for new rules, we can use right tags from now ... ?

[the remark this answers is not in the document]
Comment (AghilesAzzoug) [reply python-plugin/src/main/java/fr/greencodeinitiative/python/checks/AvoidListComprehensionInIterations.java]: Done

**Review by dedece35 — APPROVED** [review]

**Pull request merged**